apps/freeablo/faworld/player.cpp

```diff
@@ -100,34 +100,110 @@ namespace FAWorld
 
     int32_t Player::meleeDamageVs(const Actor* /*actor*/) const
     {
-        auto bonus = getItemBonus();
-        auto dmg = mWorld.mRng->randomInRange(bonus.minAttackDamage, bonus.maxAttackDamage);
-        dmg += dmg * getPercentDamageBonus() / 100;
-        dmg += getCharacterBaseDamage();
-        dmg += getDamageBonus();
-        // critical hit for warriors:
-        if (mPlayerClass == PlayerClass::warrior && mWorld.mRng->randomInRange(0, 99) < getCharacterLevel())
-            dmg *= 2;
-        return dmg;
+        const LiveActorStats& stats = mStats.getCalculatedStats();
+        int32_t damage = stats.meleeDamage;
+        damage += mWorld.mRng->randomInRange(stats.meleeDamageBonusRange.start, stats.meleeDamageBonusRange.end);
+
+        if (mPlayerClass == PlayerClass::warrior && mWorld.mRng->randomInRange(0, 99) < mStats.mLevel)
+            damage *= 2;
+
+        return damage;
     }
 
-    ItemBonus Player::getItemBonus() const { return mInventory.getTotalItemBonus(); }
+    void Player::calculateStats(LiveActorStats& stats, const ActorStats& actorStats) const
+    {
+        BaseStats charStats = actorStats.baseStats;
+
+        ItemStats itemStats;
+        mInventory.calculateItemBonuses(itemStats);
+
+        stats.baseStats.strength = charStats.strength + itemStats.baseStats.strength;
+        stats.baseStats.magic = charStats.magic + itemStats.baseStats.magic;
+        stats.baseStats.dexterity = charStats.dexterity + itemStats.baseStats.dexterity;
+        stats.baseStats.vitality = charStats.vitality + itemStats.baseStats.vitality;
+
+        stats.toHitMelee.bonus = 0;
+        stats.toHitRanged.bonus = 0;
+        stats.toHitMagic.bonus = 0;
+
+        // TODO: make sure all the following calculations should be rounded
```

Comment:
@AJenbo can you tell me how this works in vanilla? Are the values for damage, hit chance etc rounded before use, or are damage calculations done with floating point?

Reply:
This guide has the game's combat formulas:
http://www.lurkerlounge.com/diablo/jarulf/jarulf162.pdf